### PR TITLE
fixes #20139; hash types based on its path relative to its package path

### LINKED ISSUE #20139: gcc error when constructing an object that has the same name in the same file name in 2 different directories
### Example
`module1`/`defs.nim`:
```nim
type MyObj* = object
  field1*: int
```

`module2`/`defs.nim`:
```nim
type MyObj* = object
  field2*: int
```

`main.nim`:
```nim
import module1/defs as md1
import module2/defs as md2

echo md1.MyObj(field1: 1)
echo md2.MyObj(field2: 1)
```

### Current Output
gcc error, undeclared field of struct:
```
error: 'tyObject_MyObj__vPIR9aeSj2i9aDvEgKt9azXzg' {aka 'struct tyObject_MyObj__vPIR9aeSj2i9aDvEgKt9azXzg'} has no member named 'field2'; did you mean 'field1'?
  addQuoted__main_48((&result), x.field2);
```

### Expected Output
compiles successfully

### Possible Solution
maybe name mangeling doesn't happen correctly in this situation

### Additional Information
```
$ nim -v
Nim Compiler Version 1.7.1 [Windows: amd64]
Compiled at 2022-07-17
Copyright (c) 2006-2022 by Andreas Rumpf

active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20139

`hashTypeSym` is used to calculate the hash of a type symbol. To differentiate modules with the same name but in the different paths of a package, we can consider its relative path to its package path.

```nim
c &= relativePath(conf.toFullPath(s.info), conf.toFullPath(getPackageSymbol(s).info), '/')
```

### CODE REVIEW

**Comment by Araq** [comment]
Seems excessive, isn't there an easier way?

**Comment by quantimnot** [comment]
@ringabout How about including `ast.hash(ast.ItemId)` into the symbol hashes? I'm not sure of negative consequences, but it seems to work for me. I haven't run the full test suite though.

Here is what I tried:
```nim
+template includeIdHash(c: var MD5Context, s: PSym) =
+  # Include the symbol ID to disambiguate symbols with identical names and
+  # identical module names. (bug #20139)
+  c &= hash(s.itemId)

proc hashSym(c: var MD5Context, s: PSym) =
+  includeIdHash(c, s)
  if sfAnon in s.flags or s.kind == skGenericParam:
    c &= ":anon"
  else:
    var it = s
    while it != nil:
      c &= it.name.s
      c &= "."
      it = it.owner

proc hashTypeSym(c: var MD5Context, s: PSym) =
+  includeIdHash(c, s)
  if sfAnon in s.flags or s.kind == skGenericParam:
    c &= ":anon"
  else:
    var it = s
    while it != nil:
      if sfFromGeneric in it.flags and it.kind in routineKinds and
          it.typ != nil:
        hashType c, it.typ, {CoProc}
      c &= it.name.s
      c &= "."
      it = it.owner
```

**Comment by ringabout** [comment]
I'm not sure whether it is consistent enough, see also https://github.com/nim-lang/Nim/pull/20564#discussion_r995878363

**Comment by ringabout** [comment]
@Araq Could we use what @quantimnot suggested or is there another way out?

**Comment by Araq** [comment]
My own "superior" solution ended up being nearly identical... :-)

**Comment by arnetheduck** [comment]
a backport to 1.6 would be appreciated here - this completely cripples 1.6 both compile-time and runtime depending on the shape of objects

File: tests/ccgbugs/m1/defs.nim [hunk 1]
```diff
@@ -0,0 +1,2 @@
+type MyObj* = object
+  field1*: int
```
Comment (arnetheduck) [inline tests/ccgbugs/m1/defs.nim]: an additional test here would be needed, constructed the following way:
* same name
* same field names
* different field order

such an object, before this commit, will be compiled correctly but its data would be written according to the wrong object type in certain cases leading to runtime corruption

**Comment by Araq** [comment]
"path relative to its package path" is still wrong or at least inflexible. --> Wait for my change on top of this PR.

**Pull request merged**